cses concert tickets + traffic lights sol (#549)
cpinitiative/usaco-guide: solutions/CSES Traffic Lights.mdx

---
id: cses-1163
source: CSES
title: Traffic Lights
author: Danh Ta Chi Thanh
---

> In this problem, we are given an empty interval length $x$ starting from 0 to $x$, and there are $n$ points being added to the interval chronologically. We want to find the maximum lenght of the gaps between the points at each step.

Reviewer: ```suggestion
In this problem, we are given an empty interval of length $x$ spanning from 0 to $x$, and $n$ points are added to the interval chronologically. We want to find the largest gap between the points after each step.
```

> ## Solution

Reviewer: ```suggestion
## Solution 1
```

First you can read and think more about the solution here: 
[Unofficial Editorial](https://codeforces.com/blog/entry/83295) (and also here: [Sketch](https://codeforces.com/blog/entry/86096?#comment-739122))

> I will try to explain it: Let's create a set and a multiset. The set is used for storing the positions of the traffic lights, while the multiset will keep track of the "gaps" between the lights. After each addition, the multiset keeps expanding because more lights are put and your work is just to print "the length of the longest passage without traffic lights after each addition" or the max elements of that multiset which is the last element by default. Note that when placing a new traffic light on the road, that light will split the gap between two $adjacent$ lights into 2 smallers pieces, and you also have to remove the length of that gap in the multiset and add 2 new lenghts to the multiset.

Reviewer: ```suggestion
I will try to explain it: Let's create a set and a multiset. The set will store the positions of the traffic lights, while the multiset will keep track of the "gaps" between the lights. The multiset keeps expanding because more lights are added, and you just need to print the length of the longest passage without traffic lights after each addition (i.e. the max element of that multiset). This element is the last element by default.

Note that when placing a new traffic light on the road, that light will split the gap between two adjacent lights into two smaller pieces, so you also have to remove the length of that gap in the multiset and add two new lengths to the multiset.
```

> ## Code Implementation

Reviewer: Also I don't think you need this heading

Author: okay I will remove it

```cpp
#include <bits/stdc++.h>
 
using namespace std;
 
//constant initialization
const int maxn=2e5+10;
 
//variables used for the current problem
int x,n,p; set<int> lights; multiset<int> dist;
 
void solve() {
	cin >> x >> n;
	//I add the first point and the last point because when we place the very first light, 
	//it will creates two smaller intervals, which are 0-light1 and light1-x.

> 	lights.insert(0); lights.insert(x);

Reviewer: I don't think you need all these comments. Your above explanation should be sufficient

	dist.insert(x); //It is obvious that initially the longest passage has the lenght equal to x. 
	for (int i=0;i<n;++i){
		cin >> p;
		//Read the input position of the traffic light
		//Next, I find the traffic light lying in the right of the current traffic light (p) using upper_bound function
		// And I find the traffic light to the left of the traffic light p by decrement the iterator
		auto it1 = lights.lower_bound(p),it2 = it1; --it2;
		//Then I delete the lenghts of that left and right lights because it now has been split by the new light.
		dist.erase(dist.find(*it1-*it2));
		//After that, I insert the two new gaps which have just been created by adding a light between the 2 old lights.
		dist.insert(p-*it2); dist.insert(*it1-p);
		//Insert the position of the new light to the "lights" set
		lights.insert(p);
		//the answer in this current step is the last element off the multiset
		auto ans = dist.end(); --ans;
		cout << *ans << " ";
	}
}

int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	solve();
	return 0;
}
``` 

> P/S: I found another solution by reversing all the steps and calculate the maximum lenght at each step. Here is my implementation.

Reviewer: Move this line to below `## Solution 2 - ...` and take out the `P/S:`.

> P/S: I found another solution by reversing all the steps and calculate the maximum lenght at each step. Here is my implementation.

Reviewer: Also some spelling errors:

```suggestion
P/S: I found another solution by reversing all the steps and calculating the maximum length at each step. Here is my implementation.
```
